server/src/main/java/org/elasticsearch/cluster/ClusterFeatures.java

```diff
@@ -0,0 +1,266 @@
+                }
+
+                Set<String> added = new HashSet<>(newFeatures);
+                added.removeAll(prevNodeFeatures.getValue());
+                if (added.isEmpty() == false) {
+                    additions.put(prevNodeFeatures.getKey(), added);
+                }
+            }
+        }
+
+        // find any completely new nodes
+        for (var newNodeFeatures : nodeFeatures.entrySet()) {
+            if (previousState.nodeFeatures.containsKey(newNodeFeatures.getKey()) == false) {
+                additions.put(newNodeFeatures.getKey(), newNodeFeatures.getValue());
+            }
+        }
+
+        return new ClusterFeaturesDiff(deletes, removals, additions);
+    }
+
+    public static Diff<ClusterFeatures> readDiffFrom(StreamInput in) throws IOException {
+        return new ClusterFeaturesDiff(in);
+    }
+
+    private static class ClusterFeaturesDiff implements Diff<ClusterFeatures> {
+
+        private final Set<String> deletes;
+        private final Map<String, Set<String>> removals;
+        private final Map<String, Set<String>> additions;
+
+        private ClusterFeaturesDiff(Set<String> deletes, Map<String, Set<String>> removals, Map<String, Set<String>> additions) {
+            this.deletes = deletes;
+            this.removals = removals;
+            this.additions = additions;
+        }
+
+        private ClusterFeaturesDiff(StreamInput in) throws IOException {
+            deletes = in.readCollectionAsImmutableSet(StreamInput::readString);
+            removals = readCanonicalSets(in);
+            additions = readCanonicalSets(in);
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeCollection(deletes, StreamOutput::writeString);
+            writeCanonicalSets(out, removals);
+            writeCanonicalSets(out, additions);
+        }
+
+        @Override
+        public ClusterFeatures apply(ClusterFeatures part) {
+            if (deletes.isEmpty() && removals.isEmpty() && additions.isEmpty()) {
+                return part;    // nothing changing
+            }
+
+            Map<String, Set<String>> newFeatures = new HashMap<>(part.nodeFeatures);
+            deletes.forEach(newFeatures::remove);
+
+            // make sure each value is mutable when we modify it
+            for (var removes : removals.entrySet()) {
+                newFeatures.compute(removes.getKey(), (k, v) -> v instanceof HashSet ? v : new HashSet<>(v)).removeAll(removes.getValue());
+            }
+            for (var adds : additions.entrySet()) {
+                newFeatures.compute(adds.getKey(), (k, v) -> v == null ? new HashSet<>() : v instanceof HashSet ? v : new HashSet<>(v))
+                    .addAll(adds.getValue());
+            }
+
+            return new ClusterFeatures(newFeatures);
+        }
+    }
+
+    @Override
+    public Iterator<? extends ToXContent> toXContentChunked(ToXContent.Params params) {
```

Comment:
Do we need to be able to serialize to XContent? 
If it is for rest action(s) I don't think we have decided (yet) if we want to publicly expose features via a REST API. I naively thought yes, we would have them exposed in an existing API (e.g. in `GET /_cluster/state/` or `GET _nodes`, alongside `version`) or through a new ad-hoc one, but a comment by @rjernst made me think twice about this. Are we exposing an implementation detail? Can/should we avoid it?

(if we do expose features through a REST API, it would be easier to work with features in rest tests, but I'm not sure that creating an API for testing purposes only is the way to go)

Reply:
We do - this is serialized as part of `ClusterState`

Reply:
We need some way to diagnose & debug features that may not be used on a cluster - this is the standard way to get that info

Reply:
_Everything_ the cluster state API returns is an implementation detail :grin: It even says so in [its docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/cluster-state.html):

> Returns an internal representation of the cluster state for debugging or diagnostic purposes.
> ...
> **WARNING**: The response is a representation of an internal data structure. Its format is not subject to the same compatibility guarantees as other more stable APIs and may change from version to version. Do not query this API using external monitoring tools. Instead, obtain the information you require using other more stable [cluster APIs](https://www.elastic.co/guide/en/elasticsearch/reference/current/cluster.html).

Reply:
TIL, thank you for clarifying that! This means we could use it for rest tests (features check for the cluster under tests, on the rest-client (java or yaml) side), right?

Reply:
Ehm I would say those usages deserve to use a dedicated API, although we could technically get away with using the cluster state API for them too.

Reply:
(I do not want to derail the conversation from this PR) Today most of these usages are using `GET _nodes` (to get NodeInfo.versions, and infer features from them) with some using the CAT API instead; we started to discuss some ideas and options on how to replace them with features checks - a dedicated REST API was one of them, but not the only option. That's why I asked my original question, wrongly assuming we already committed in that direction, on why we need XContent (which has been answered fully).